packages/zoltan2/core/src/algorithms/color/Zoltan2_AlgHybrid2GL.hpp

```diff
@@ -0,0 +1,1532 @@
+    RCP<const Teuchos::Comm<int> > comm;
+    bool verbose;
+
+    
+  private:
+    //This function constructs the second ghost layer.
+    //It proceeds in rounds in order to keep messages
+    //under MPI's maximum send size.
+    //
+    //OUTPUT ARGS
+    //  ownedPlusGhosts holds the initial list of GIDs for all owned vertices.
+    //                  This function changes the order of the ghost GIDs in
+    //                  order to more easily construct the second ghost layer.
+    //
+    //  adjs_2GL holds the second ghost layer CSR adjacencies
+    //
+    //  offsets_2GL holds CSR offsets into the second ghost layer adjacencies.
+    //
+    //INPUT ARGS
+    //  owners holds the owning proc for a given vertex, indexed by local ID.
+    //
+    //  adjs is the CSR adjacencies of the local graph with a single ghost layer.
+    //
+    //  offsets is the CSR offsets of the local graph with a single ghost layer.
+    //
+    //  mapOwned translates from Owned GID to LID. We only need this translation
+    //           for owned vertices.
+    //
+    void constructSecondGhostLayer(std::vector<gno_t>& ownedPlusGhosts,
+                                   const std::vector<int>& owners,
+                                   ArrayView<const gno_t> adjs,
+                                   ArrayView<const offset_t> offsets,
+                                   RCP<const map_t> mapOwned,
+                                   std::vector< gno_t>& adjs_2GL,
+                                   std::vector< offset_t>& offsets_2GL) {
+      std::vector<int> sendcounts(comm->getSize(),0);
+      std::vector<gno_t> sdispls(comm->getSize()+1,0);
+      //loop through owners, count how many vertices we'll send to each processor
+      if(verbose) std::cout<<comm->getRank()<<": building sendcounts\n";
+      for(size_t i = 0; i < owners.size(); i++){
+        if(owners[i] != comm->getRank()&& owners[i] !=-1) sendcounts[owners[i]]++;
+      }
+      //construct sdispls (for building sendbuf), and sum the total sendcount
+      if(verbose) std::cout<<comm->getRank()<<": building sdispls\n";
+      gno_t sendcount = 0;
+      for(int i = 1; i < comm->getSize()+1; i++){
+        sdispls[i] = sdispls[i-1] + sendcounts[i-1];
+        sendcount += sendcounts[i-1];
+      }
+
+      if(verbose) std::cout<<comm->getRank()<<": building idx\n";
+      std::vector<gno_t> idx(comm->getSize(),0);
+      for(int i = 0; i < comm->getSize(); i++){
+        idx[i]=sdispls[i];
+      }
+      //construct sendbuf to send GIDs to owning processes
+      if(verbose) std::cout<<comm->getRank()<<": building sendbuf\n";
+
+      std::vector<gno_t> sendbuf(sendcount,0);
+      for(size_t i = offsets.size()-1; i < owners.size(); i++){
+        if(owners[i] != comm->getRank() && owners[i] != -1){
+          sendbuf[idx[owners[i]]++] = ownedPlusGhosts[i];
+        }
+      }
+      //remap the GIDs so we receive the adjacencies in the same order as the current processes LIDs
```

Comment:
More detail here would be helpful.  What was the order going in?  What is the order coming out?
KDD:  Finished to here; will resume later.